### PR TITLE
dialects: (stim) add qubit attribute and qubit coordinate attribute

### PR DESCRIPTION
**Pull request by kimxworrall**
Add a base attribute to the stim dialect - StimAttr which is StimPrintable so that all implementing attributes must implement print_stim.

Add two initial attributes - QubitAttr to indicate qubits, and QubitCoordAttr which provides a pair of a coordinate of a physical qubit to a QubitAttr

### CODE REVIEW

File: xdsl/dialects/stim/ops.py [hunk 1]
```diff
@@ -1,24 +1,119 @@
...
 
-from .stim_printer_parser import StimPrintable, StimPrinter
 
+class StimAttr(StimPrintable):
+    "Base Stim attribute"
 
-class StimOp(IRDLOperation, ABC):
-    def print_stim(self, printer: StimPrinter) -> None:
-        raise (PyRDLOpDefinitionError("print_stim not implemented!"))
+    name = "stim.attr"
```
Comment (superlopuh) [inline xdsl/dialects/stim/ops.py:21]: What is the purpose of this class?

File: xdsl/dialects/stim/ops.py [hunk 2]
```diff
@@ -1,24 +1,119 @@
...
+        if not isinstance(qubit, IntAttr):
+            qubit = IntAttr(qubit)
+        super().__init__(parameters=[qubit])
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[IntAttr]:
+        parser.parse_punctuation("<")
+        qubit = parser.parse_integer(allow_negative=False, allow_boolean=False)
+        parser.parse_punctuation(">")
+        return [IntAttr(qubit)]
```
Comment (superlopuh) [inline xdsl/dialects/stim/ops.py:44]: nit:
```suggestion
        with parser.in_angle_brackets():
            qubit = parser.parse_integer(allow_negative=False, allow_boolean=False)
            return [IntAttr(qubit)]
```

File: xdsl/dialects/stim/ops.py [hunk 3]
```diff
@@ -1,24 +1,119 @@
...
+
+    def print_parameters(self, printer: Printer) -> None:
+        with printer.in_angle_brackets():
+            printer.print(self.qubit.data)
+
+    def print_stim(self, printer: StimPrinter):
+        printer.print_string(self.qubit.data.__str__())
```
Comment (superlopuh) [inline xdsl/dialects/stim/ops.py:51]:
```suggestion
        printer.print_string(f"{self.qubit.data}")
```

File: xdsl/dialects/stim/stim_printer_parser.py [hunk 4]
```diff
@@ -19,6 +23,32 @@ def in_braces(self):
         finally:
             self.print_string("}")
 
+    @contextmanager
+    def in_parens(self):
+        self.print_string("(")
+        try:
+            yield
+        finally:
+            self.print_string(")")
```
Comment (superlopuh) [inline xdsl/dialects/stim/stim_printer_parser.py:32]: This was never required, I'm not sure why we have this pattern
```suggestion
        yield
        self.print_string(")")
```

**Review by superlopuh — COMMENTED** [review]
This PR adds some code and infrastructure for stim printing, but never tests it. We should either add the printing code in a future PR or add the tests here.

File: tests/dialects/stim/test_stim.py
Comment (superlopuh) [inline tests/dialects/stim/test_stim.py:1]: could you please move the testing for attribute syntax to a filecheck file, something like `tests/filecheck/dialects/stablehlo/attrs.mlir`?

File: xdsl/dialects/stim/ops.py (hunk 1 above)
Comment (kimxworrall) [reply xdsl/dialects/stim/ops.py:21]: Everything in this file is stim printable - in mlir the convention seems to be making a base class for the attributes and operations - is it okay to not do that here?

Comment (superlopuh) [reply xdsl/dialects/stim/ops.py:21]: We generally try to do it only when necessary. In this case, the fact that it has a name, and is a base class without inheriting from ABC are both a little confusing. I would just remove it for the time being.

**Review by superlopuh — APPROVED** [review]

**Comment by superlopuh** [comment]
Let's merge this?

**Pull request merged**